## 10. Error Handling with Exceptions

1. <a href = "#1"> Basic exceptions
2. <a href = "#2">Exception arguments
3. <a href = "#3">Termination vs. resumption
4. <a href = "#4">Creating your own exceptions
5. <a href = "#5">Exceptions and logging
6. <a href = "#6">
7. <a href = "#7">

<h3 id = "1"> Basic exceptions </h3>
An *exceptional condition* is a problem that prevents the continuation of the current method or scope. With an exceptional condition, you cannot continue processing because you don’t have the information necessary to deal with the problem *in the current context*. All you can do is jump out of the current context and relegate that problem to a higher context. This is what happens when you throw an exception.

When you throw an exception, several things happen. First, the exception object is created in the same way that any Java object is created: on the heap, with **new**. Then the current path of execution (the one you couldn’t continue) is stopped and the reference for the exception object is ejected from the current context. At this point the exception-handling mechanism takes over and begins to look for an appropriate place to continue executing the program. This appropriate place is the *exception handler*, whose job is to recover from the problem so the program can either try another tack or just continue.

You can send information about the error into a larger context by creating an object representing your information and "throwing" it out of your current context. This is called *throwing an exception*. Here’s what it looks like:

```java
if(t == null)
    throw new NullPointerException();
```

This throws the exception, which allows you—in the current context—to abdicate responsibility for thinking about the issue further. It’s just magically handled somewhere else. Precisely where will be shown shortly.

Exceptions allow you to (if nothing else) force the program to stop and tell you what went wrong, or (ideally) force the program to deal with the problem and return to a stable state.

<h3 id = "2">Exception arguments</h3>
As with any object in Java, you always create exceptions on the heap using new, which allocates storage and calls a constructor. There are two constructors in all standard exceptions: The first is the default constructor, and the second takes a string argument so that you can place pertinent information in the exception:

```java
throw new NullPointerException("t = null");
```

Simplistic way to think about exception handling is as a different kind of return mechanism, although you get into trouble if you take that analogy too far. You can also exit from ordinary scopes by throwing an exception. In either case, an exception object is returned, and the method or scope exits.

In addition, you can throw any type of **Throwable**, which is the exception root class. Typically, you’ll throw a different class of exception for each different type of error. The information about the error is represented both inside the exception object and implicitly in the name of the exception class, so someone in the bigger context can figure out what to do with your exception.

<h3 id = "3">Termination vs. resumption</h3>
Java supportst *termination*, in which you assume that the error is so critical that there’s no way to get back to where the exception occurred. Whoever threw the exception decided that there was no way to salvage the situation, and they don’t *want* to come back.

The alternative is called *resumption*. It means that the exception handler is expected to do something to rectify the situation, and then the faulting method is retried, presuming success the second time. 

If you want resumption-like behavior in Java, don’t throw an exception when you encounter an error. Instead, call 2a method that fixes the problem. Alternatively, place your try block inside a **while** loop that keeps reentering the **try** block until the result is satisfactory.

<h3 id = "4">Creating your own exceptions</h3>

The most trivial way to create a new type of exception is just to let the compiler create the default constructor for you, so it requires almost no code at all:

```java
class SimpleException extends Exception {}
public class InheritingExceptions {
    public void f() throws SimpleException {
        System.out.println("Throw SimpleException from f()");
        throw new SimpleException();
    }
    public static void main(String[] args) {
        InheritingExceptions sed = new InheritingExceptions();
        try {
            sed.f();
        } catch(SimpleException e) {
            System.out.println("Caught it!");
        }
    }
} /* Output:
Throw SimpleException from f()
Caught it!
*///:~
```

However, you may want to send error output to the standard error stream by writing to **System.err**. This is usually a better place to send error information than **System.out**, which may be redirected. If you send output to **System.err**, it will not be redirected along with **System.out** so the user is more likely to notice it. You can also create an exception class that has a constructor with a String argument:

```java
class MyException extends Exception {
    public MyException() {}
    public MyException(String msg) { super(msg); }
}
public class FullConstructors {
    public static void f() throws MyException {
        System.out.println("Throwing MyException from f()");
        throw new MyException();
    }
    public static void g() throws MyException {
        System.out.println("Throwing MyException from g()");
        throw new MyException("Originated in g()");
    }
    public static void main(String[] args) {
        try {
            f();
        } catch(MyException e) {
            e.printStackTrace(System.out);
        }
        try {
            g();
        } catch(MyException e) {
            e.printStackTrace(System.out);
        }
    }
} /* Output:
Throwing MyException from f()
MyException
at FullConstructors.f(FullConstructors.java:11)
at FullConstructors.main(FullConstructors.java:19)
Throwing MyException from g()
MyException: Originated in g()
at FullConstructors.g(FullConstructors.java:15)
at FullConstructors.main(FullConstructors.java:24)
*///:~
```

In the handlers, one of the **Throwable** (from which **Exception** is inherited) methods is called: **printStackTrace( )**. As you can see from the output, this produces information about the sequence of methods that were called to get to the point where the exception happened. Here, the information is sent to System.out, and automatically captured and displayed in the output. However, if you call the default version:

```java
e.printStackTrace();
```

the information goes to the standard error stream.

<h3 id = "5">Exceptions and logging</h3>

You may also want to log the output using the java.util.logging facility.

```java
import java.util.logging.*;
import java.io.*;
class LoggingException extends Exception {
    private static Logger logger =
            Logger.getLogger("LoggingException");
    public LoggingException() {
        StringWriter trace = new StringWriter();
        printStackTrace(new PrintWriter(trace));
        logger.severe(trace.toString());
    }
}
public class LoggingExceptions {
    public static void main(String[] args) {
        try {
            throw new LoggingException();
        } catch(LoggingException e) {
            System.err.println("Caught " + e);
        }
        try {
            throw new LoggingException();
        } catch(LoggingException e) {
            System.err.println("Caught " + e);
        }
    }
} /* Output: (85% match)
Aug 30, 2005 4:02:31 PM LoggingException <init>
SEVERE: LoggingException
at LoggingExceptions.main(LoggingExceptions.java:19)
Caught LoggingException
Aug 30, 2005 4:02:31 PM LoggingException <init>
SEVERE: LoggingException
at LoggingExceptions.main(LoggingExceptions.java:24)
Caught LoggingException
*///:~
```

The **static** **Logger.getLogger( )** method creates a **Logger** object associated with the **String** argument (usually the name of the package and class that the errors are about) which sends its output to **System.err**. The easiest way to write to a **Logger** is just to call the method associated with the level of logging message; here, **severe( )** is used. To produce the **String** for the logging message, we’d like to have the stack trace where the exception is thrown, but **printStackTrace( )** doesn’t produce a **String** by default. To get a **String**, we need to use the overloaded **printStackTrace( )** that takes a **java.io.PrintWriter** object as an argument. If we hand the **Print** **Writer** constructor a **java.io.StringWriter** object, the output can be extracted as a **String** by calling **toString( )**.

